src/test/ui/nll/type-check-pointer-comparisons.rs

```diff
@@ -21,13 +21,13 @@ fn compare_fn_ptr<'a, 'b, 'c>(f: fn(&'c mut &'a i32), g: fn(&'c mut &'b i32)) {
 }
 
 fn compare_hr_fn_ptr<'a>(f: fn(&'a i32), g: fn(&i32)) {
-    // Ideally this should compile with the operands swapped as well, but HIR
-    // type checking prevents it (and stops compilation) for now.
-    f == g; // OK
+    f == g;
+    //~^ ERROR higher-ranked subtype error
```

Comment:
Hmm, I guess the code here isn't doing what I thought it was
https://github.com/rust-lang/rust/blob/96d77f0e5f103612d62b85938aacfb33f5768433/src/librustc_mir/borrow_check/type_check/mod.rs#L2302-L2333

I think I was expecting this to implicitly do `self.infcx.resolve_vars_if_possible(common_ty)` in the second `sub_types` call.

Reply:
I don't think that would have any effect. What happens in the first call is that we instantiate `common_ty` to a version of `ty_left`, but with distinct region variables (which are related to the regions in `ty_left`). It sounds like you were expecting `common_ty` and `ty_left` to get unified?

Reply:
No, I was expecting `common_ty` to be a generalized version of `ty_left`. I was then expecting the next call to relate the same generalized type to `ty_right`, but it is instead equating `common_ty` to a generalized version of `ty_right`.

In this specific case I was expecting `common_ty` to be instantiated as `fn(&'? i32)` and then be constrained to be a supertype of `fn(&'a i32)` and `fn(&i32)`. Instead it's instantiated as both `fn(&'? i32)` and `fn(&i32)`, which is an error.

Reply:
I can't tell if you're suggesting a change here or not =)

Reply:
That said, I think I was expecting the same as you. Also, re-reading the code, I'm a bit.. surprised by it. It seems like there's a baked-in assumption about the sort of types that we'll apply the comparison operator to that probably ought to be checked (i.e., it seems like we're assuming that in all other cases, the values are trivially of the same type, I guess?).

Still, I'm not entirely sure how you are deriving your conclusions. I think I expected:

* `common_ty` gets instantiated to `fn(&'? i32)` (where `'a: '?`, unless I got something backward)
* we then require that `fn(&'? i32) <: fn(&i32)`, which yields a requirement that `for<'b> { 'b: '? }`, which we cannot satisfy (though, prior to this PR, we could)
* hence we get a higher-ranked error

In particular I don't really know what this means:

> Instead it's instantiated as both `fn(&'? i32)` and `fn(&i32)`, which is an error.

How can a single variable be instantiated as two types? Maybe we mean something different by 'instantiated'.

Reply:
Looking at the code again, I was wrong what the problem was. The types are being related the opposite way to they actually should be.

`common_ty` is the type that the comparison is being done at, so each of the operand should be a subtype of `common_ty`, that is `common_ty` should be a common *super*-type.

> It seems like there's a baked-in assumption about the sort of types that we'll apply the comparison operator to that probably ought to be checked (i.e., it seems like we're assuming that in all other cases, the values are trivially of the same type, I guess?).

Yes, this should probably be changed to match the remaining types where `BinOp::Eq` is allowed (ints, `bool`, `char`) and assert equality I guess this can be done separately.

Reply:
@matthewjasper good point, I suppose we do want a common super-type, not a common sub-type! I'll push a commit making this change.

Reply:
Update: it wasn't easy to change the order. This is because, I think, the `relate_tys` code still has some hard-coded assumptions that assume that the super-type is not an uninstantiated variable, or something like that. This may have been what led to the code being written this way in the first place.